pkg/reconciler/autoscaling/hpa/resources/keda.go

```diff
@@ -0,0 +1,112 @@
+package resources
+
+import (
+	"fmt"
+	"math"
+
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/ptr"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	autoscalingv1alpha1 "knative.dev/serving/pkg/apis/autoscaling/v1alpha1"
+	"knative.dev/serving/pkg/autoscaler/config/autoscalerconfig"
+)
+
+// MakeScaledObject creates an ScaledObject KEDA resource from a PA resource.
+func MakeScaledObject(pa *autoscalingv1alpha1.PodAutoscaler, config *autoscalerconfig.Config) *v1alpha1.ScaledObject {
+	min, max := pa.ScaleBounds(config)
+	if max == 0 {
+		max = math.MaxInt32 // default to no limit
+	}
+
+	sO := v1alpha1.ScaledObject{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            pa.Name,
+			Namespace:       pa.Namespace,
+			Labels:          pa.Labels,
+			Annotations:     pa.Annotations,
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(pa)},
+		},
+		Spec: v1alpha1.ScaledObjectSpec{
+			ScaleTargetRef: &v1alpha1.ScaleTarget{
+				Name: pa.Name + "-deployment",
+			},
+			Advanced: &v1alpha1.AdvancedConfig{
+				HorizontalPodAutoscalerConfig: &v1alpha1.HorizontalPodAutoscalerConfig{
+					Name: pa.Name,
+				},
+			},
+			MaxReplicaCount: ptr.Int32(max),
+		},
+	}
+
+	if min > 0 {
+		sO.Spec.MinReplicaCount = ptr.Int32(min)
+	}
+
+	if target, ok := pa.Target(); ok {
+		switch pa.Metric() {
+		case autoscaling.CPU:
+			sO.Spec.Triggers = []v1alpha1.ScaleTriggers{
+				{
+					Type:       "cpu",
+					MetricType: autoscalingv2.UtilizationMetricType,
+					Metadata:   map[string]string{"value": string(int32(math.Ceil(target)))},
+				},
+			}
+		case autoscaling.Memory:
+			memory := resource.NewQuantity(int64(target)*1024*1024, resource.BinarySI)
+			sO.Spec.Triggers = []v1alpha1.ScaleTriggers{
+				{
+					Type:       "memory",
+					MetricType: autoscalingv2.AverageValueMetricType,
+					Metadata:   map[string]string{"value": memory.String()},
+				},
+			}
+		default:
+			if target, ok := pa.Target(); ok {
+				targetQuantity := resource.NewQuantity(int64(target), resource.DecimalSI)
+				var query, address string
+				if v, ok := pa.Annotations["autoscaling.knative.dev/query"]; ok {
```

Comment:
Public API? Probably should be defined somewhere as public const?

Reply:
Yes it should be I will change.